File: core/java/src/main/java/hara/lang/protocol/IWorkHost.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Process-owned host that admits work and resolves live run handles. */
@HaraProtocolBinding(
    namespace = "std.protocol.iworkhost",
    name = "IWorkHost",
    parents = {"IComponent"},
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IWorkHost extends IComponent {
  @HaraMethod(value = "work-submit", arity = 4)
  IWorkRun workSubmit(Object work, Object input, Object options);

  @HaraMethod(value = "work-resolve", arity = 2)
  IWorkRun workResolve(Object reference);
}

File: core/rust/src/lang/protocol/iworkhost.rs
use super::IComponent;
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iworkhost",
    name = "IWorkHost",
    parents = ["IComponent"],
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IWorkHost: IComponent {
    type Work;
    type Input;
    type Options;
    type Reference;
    type Run;

    #[hara_method(value = "work-submit", arity = 4)]
    fn work_submit(
        &self,
        work: Self::Work,
        input: Self::Input,
        options: Self::Options,
    ) -> Self::Run;

    #[hara_method(value = "work-resolve", arity = 2)]
    fn work_resolve(&self, reference: Self::Reference) -> Self::Run;
}
